From 038520564d54c59d66172d4c2a1bf512109d3532 Mon Sep 17 00:00:00 2001
From: Benjamin Walsh <benjamin.walsh@windriver.com>
Date: Fri, 24 Sep 2010 16:09:42 -0400
Subject: [PATCH 169/233] kexec/ppc32: purgatory expect r3 to contain boot CPU#

Purgatory will get the value of r3 and force it in the DT
passed to the second kernel. If the boot CPU is not the same
as that value, the boot of the second kernel will be
interrupted and the kernel will oops.

Hmm, how come we've never been bitten by this one ? It only
caused a problem on P4080 crash reboot.

Signed-off-by: Benjamin Walsh <benjamin.walsh@windriver.com>
---
 arch/powerpc/kernel/misc_32.S |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/kernel/misc_32.S b/arch/powerpc/kernel/misc_32.S
index 1d0522c..1e24d1e 100644
--- a/arch/powerpc/kernel/misc_32.S
+++ b/arch/powerpc/kernel/misc_32.S
@@ -957,6 +957,9 @@ all_cpus:
 	addi	r25, r4, relocate_new_kernel_spin_addr - relocate_new_kernel
 #endif
 
+	/* purgatory expects r3 to contain the boot CPU: make it happy */
+	mfspr	r3, SPRN_PIR /* who are we ? */
+
 	/* jump to the entry point, usually the purgatory */
 	mtlr	r5
 	blrl
-- 
1.7.0.4

